spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java

```diff
@@ -0,0 +1,552 @@
+
+    ExecutorService rewriteService = rewriteService();
+
+    ConcurrentLinkedQueue<String> rewrittenIDs = new ConcurrentLinkedQueue<>();
+    ConcurrentHashMap<String, Pair<FileGroupInfo, FileGroupRewriteResult>> results = new ConcurrentHashMap<>();
+
+    Tasks.Builder<FileGroup> rewriteTaskBuilder = Tasks.foreach(groupStream)
+        .executeWith(rewriteService)
+        .stopOnFailure()
+        .noRetry()
+        .onFailure((fileGroup, exception) -> {
+          LOG.error("Failure during rewrite process for group {}", fileGroup.info, exception);
+        });
+
+    try {
+      rewriteTaskBuilder
+          .run(fileGroup -> rewriteFiles(ctx, fileGroup, rewrittenIDs, results));
+    } catch (Exception e) {
+      // At least one rewrite group failed, clean up all completed rewrites
+      LOG.error("Cannot complete rewrite, {} is not enabled and one of the file set groups failed to " +
+          "be rewritten. This error occurred during the writing of new files, not during the commit process. This" +
+          "indicates something is wrong that doesn't involve conflicts with other Iceberg operations. Enabling" +
+          "{} may help in this case but the root cause should be investigated. Cleaning up {} groups which finished " +
+          "being written.", PARTIAL_PROGRESS_ENABLED, PARTIAL_PROGRESS_ENABLED, rewrittenIDs.size(), e);
+
+      Tasks.foreach(rewrittenIDs)
+          .suppressFailureWhenFinished()
+          .run(this::abortFileGroup);
+      throw e;
+    } finally {
+      rewriteService.shutdown();
+    }
+
+    try {
+      commitOrClean(ImmutableSet.copyOf(rewrittenIDs));
+    } catch (ValidationException | CommitFailedException e) {
+      String errorMessage = String.format(
+          "Cannot commit rewrite because of a ValidationException or CommitFailedException. This usually means that" +
+              "this rewrite has conflicted with another concurrent Iceberg operation. To reduce the likelihood of" +
+              "conflicts, set %s which will break up the rewrite into multiple smaller commits controlled by %s." +
+              "Separate smaller rewrite commits can succeed independently while any commits that conflict with" +
+              "another Iceberg operation will be ignored. This mode will create additional snapshots in the table " +
+              "history, one for each commit.",
+          PARTIAL_PROGRESS_ENABLED, PARTIAL_PROGRESS_MAX_COMMITS);
+      throw new RuntimeException(errorMessage, e);
+    } catch (Exception e) {
+      throw e;
+    }
+
+    return new Result(results.values().stream().collect(Collectors.toMap(Pair::first, Pair::second)));
+  }
+
+  private Result doExecuteWithPartialProgress(RewriteExecutionContext ctx, Stream<FileGroup> groupStream) {
+
+    ExecutorService rewriteService = rewriteService();
+
+    ExecutorService committerService =  Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder()
```

Comment:
nit: formatting

Reply:
This is the style used in our other executor creation?
I'm not sure what the preferred formatting is here - for example from ThreadPools.java

```  private static final ExecutorService WORKER_POOL = MoreExecutors.getExitingExecutorService(
      (ThreadPoolExecutor) Executors.newFixedThreadPool(
          WORKER_THREAD_POOL_SIZE,
          new ThreadFactoryBuilder()
              .setDaemon(true)
              .setNameFormat("iceberg-worker-pool-%d")
              .build()));
```

Reply:
The current one makes sense.